# AirBnB using K-Nearest Neighbors

AirBnB is a marketplace for short term rentals that allows you to list part or all of your living space for others to rent. You can rent everything from a room in an apartment to your entire house on AirBnB. Because most of the listings are on a short-term basis, AirBnB has grown to become a popular alternative to hotels. The company itself has grown from it's founding in 2008 to a 30 billion dollar valuation in 2016 and is currently worth more than any hotel chain in the world.

One challenge that hosts looking to rent their living space face is determining the optimal nightly rent price. In many areas, renters are presented with a good selection of listings and can filter on criteria like price, number of bedrooms, room type and more. Since AirBnB is a marketplace, the amount a host can charge on a nightly basis is closely linked to the dynamics of the marketplace. Here's a screenshot of the search experience on AirBnB:

<img src="1.jpg">

As a host, if we try to charge above market price for a living space we'd like to rent, then renters will select more affordable alternatives which are similar to ours. If we set our nightly rent price too low, we'll miss out on potential revenue.

One strategy we could use is to:

* find a few listings that are similar to ours,
* average the listed price for the ones most similar to ours,
* set our listing price to this calculated average price.


In this mission, we'll explore a specific machine learning technique called **k-nearest neighbors**, which mirrors the strategy we just described. Before we dive further into machine learning and k-nearest neighbors, let's get familiar with the dataset we'll be working with.

# Dataset

While AirBnB doesn't release any data on the listings in their marketplace, a separate group named Inside AirBnB has extracted data on a sample of the listings for many of the major cities on the website. In this post, we'll be working with their dataset from October 3, 2015 on the listings from Washington, D.C., the capital of the United States. Each row in the dataset is a specific listing that's available for renting on AirBnB in the Washington, D.C. area.

To make the dataset less cumbersome to work with, we've removed many of the columns in the original dataset and renamed the file to dc_airbnb.csv. Here are the columns we kept:

* host_response_rate: the response rate of the host
* host_acceptance_rate: number of requests to the host that convert to rentals
* host_listings_count: number of other listings the host has
* latitude: latitude dimension of the geographic coordinates
* longitude: longitude part of the coordinates
* city: the city the living space resides
* zipcode: the zip code the living space resides
* state: the state the living space resides
* accommodates: the number of guests the rental can accommodate
* room_type: the type of living space (Private room, Shared room or Entire home/apt
* bedrooms: number of bedrooms included in the rental
* bathrooms: number of bathrooms included in the rental
* beds: number of beds included in the rental
* price: nightly price for the rental
* cleaning_fee: additional fee used for cleaning the living space after the guest leaves
* security_deposit: refundable security deposit, in case of damages
* minimum_nights: minimum number of nights a guest can stay for the rental
* maximum_nights: maximum number of nights a guest can stay for the rental
* number_of_reviews: number of reviews that previous guests have left

In [4]:
import pandas as pd
import numpy as np


In [5]:
dc_listings = pd.read_csv('listings.csv')
dc_listings.head()

id                           listing_url       scrape_id last_scraped  \
0  7087327  https://www.airbnb.com/rooms/7087327  20151002231825   2015-10-03   
1   975833   https://www.airbnb.com/rooms/975833  20151002231825   2015-10-03   
2  8249488  https://www.airbnb.com/rooms/8249488  20151002231825   2015-10-03   
3  8409022  https://www.airbnb.com/rooms/8409022  20151002231825   2015-10-03   
4  8411173  https://www.airbnb.com/rooms/8411173  20151002231825   2015-10-03   

                                 name  \
0  Historic DC Condo-Walk to Capitol!   
1     Spacious Capitol Hill Townhouse   
2    Spacious/private room for single   
3    A wonderful bedroom with library   
4              Downtown Silver Spring   

                                             summary  \
0  Professional pictures coming soon! Welcome to ...   
1                                                NaN   
2  This is an ideal room for a single traveler th...   
3  Prime location right on the Potomac River in W...   
4  Hi travellers! I live in this peaceful spot, b...   

                                               space  \
0                                                NaN   
1  Beautifully renovated Capitol Hill townhouse. ...   
2                                                NaN   
3                                                NaN   
4  This is a 750 sq ft 1 bedroom 1 bathroom.  Whi...   

                                         description experiences_offered  \
0  Professional pictures coming soon! Welcome to ...                none   
1  Beautifully renovated Capitol Hill townhouse. ...                none   
2  This is an ideal room for a single traveler th...                none   
3  Prime location right on the Potomac River in W...                none   
4  Hi travellers! I live in this peaceful spot, b...                none   

                               neighborhood_overview  ... review_scores_value  \
0                                                NaN  ...                 NaN   
1                                                NaN  ...                 9.0   
2                                                NaN  ...                 NaN   
3                                                NaN  ...                 NaN   
4  Silver Spring is booming.  You can walk to a n...  ...                 NaN   

  requires_license license                jurisdiction_names instant_bookable  \
0                f     NaN  DISTRICT OF COLUMBIA, WASHINGTON                f   
1                f     NaN  DISTRICT OF COLUMBIA, WASHINGTON                f   
2                f     NaN                               NaN                f   
3                f     NaN  DISTRICT OF COLUMBIA, WASHINGTON                f   
4                f     NaN                               NaN                f   

  cancellation_policy  require_guest_profile_picture  \
0            flexible                              f   
1              strict                              f   
2            flexible                              f   
3            flexible                              f   
4            flexible                              f   

  require_guest_phone_verification calculated_host_listings_count  \
0                                f                             18   
1                                f                              1   
2                                f                              1   
3                                f                              1   
4                                f                              1   

  reviews_per_month  
0               NaN  
1              2.11  
2              1.00  
3               NaN  
4               NaN  

[5 rows x 92 columns]

The similarity metric works by comparing a fixed set of numerical **features**, another word for attributes, between 2 **observations**, or living spaces in our case. When trying to predict a continuous value, like price, the main similarity metric that's used is **Euclidean distance**. Here's the general formula for Euclidean distance:

<img src='Euclidean distance.jpg'>

where 
q1 to qn represent the feature values for one observation and 
p1 to pn represent the feature values for the other observation. Here's a diagram that breaks down the Euclidean distance between the first 2 observations in the dataset using only the host_listings_count, accommodates, bedrooms, bathrooms, and beds columns:

<img src='Eu2.jpg'>


The living space that we want to rent can accommodate 3 people. Let's first calculate the distance, using just the accommodates feature, between the first living space in the dataset and our own.

The Euclidean distance between the first row in the dc_listings Dataframe and our own living space is 1. How do we know if this is high or low? If you look at the Euclidean distance equation itself, the lowest value you can achieve is 0. This happens when the value for the feature is exactly the same for both observations you're comparing. 

If p1=q1, then d=|q1−p1| which results in d=0. The closer to 0 the distance the more similar the living spaces are.

If we wanted to calculate the Euclidean distance between each living space in the dataset and a living space that accommodates 8 people, here's a preview of what that would look like.

<img src = "Dist_8list.jpg">

In [6]:
new_listing = 3
dc_listings['distance'] = dc_listings['accommodates'].apply(lambda x: np.abs(x-new_listing))
print(dc_listings['distance'].value_counts())

1     2294
2      503
0      461
3      279
5       73
4       35
7       22
6       17
9       12
13       8
8        7
12       6
11       4
10       2
Name: distance, dtype: int64


It looks like there are quite a few, 461 to be precise, living spaces that can accommodate 3 people just like ours. This means the 5 "nearest neighbors" we select after sorting all will have a distance value of 0. If we sort by the distance column and then just select the first 5 living spaces, we would be biasing the result to the ordering of the dataset.

In [7]:
print(dc_listings[dc_listings["distance"] == 0]["accommodates"])

26      3
34      3
36      3
40      3
44      3
       ..
3675    3
3697    3
3707    3
3714    3
3722    3
Name: accommodates, Length: 461, dtype: int64


Let's instead randomize the ordering of the dataset and then sort the Dataframe by the distance column. This way, all of the living spaces with the same number of bedrooms will still be at the top of the Dataframe but will be in random order across the first 461 rows.

**np.random.permutation()** function to return a NumPy array of shuffled index values

In [8]:
np.random.seed(1)

dc_listings = dc_listings.loc[np.random.permutation(len(dc_listings))]
dc_listings = dc_listings.sort_values('distance')
print(dc_listings.iloc[0:10]['price'])

577     $185.00
2166    $180.00
3631    $175.00
71      $128.00
1011    $115.00
380     $219.00
943     $125.00
3107    $250.00
1499     $94.00
625     $150.00
Name: price, dtype: object


Before we can select the 5 most similar living spaces and compute the average price, we need to clean the price column. Right now, the price column contains comma characters (,) and dollar sign characters and is formatted as a text column instead of a numeric one. We need to remove these values and convert the entire column to the float datatype. Then, we can calculate the average price.

In [11]:
dc_listings['price'] = dc_listings['price'].str.replace(',','')
dc_listings['price'] = dc_listings['price'].str.replace('$','')
dc_listings['price'] = dc_listings['price'].astype('float')

In [12]:
dc_listings['price']

577      185.0
2166     180.0
3631     175.0
71       128.0
1011     115.0
         ...  
1596     299.0
1818      10.0
1402    1200.0
763     1000.0
1224     499.0
Name: price, Length: 3723, dtype: float64

In [13]:
mean_price = dc_listings.iloc[0:5]['price'].mean()
print(mean_price)

156.6


This is the first prediction! Based on the average price of other listings that accommdate 3 people, we should charge 156.6 dollars per night for a guest to stay at our living space.



In [20]:

def predict_price(new_listing):
    temp_df = dc_listings.copy()
    temp_df['distance'] = temp_df['accommodates'].apply(lambda x:np.abs(x - new_listing))
    #temp_df = temp_df.loc(np.random.permutation(len(temp_df)))
    temp_df = temp_df.sort_values('distance')
    mean_price = temp_df.iloc[0:5]['price'].mean()
    return mean_price

In [21]:
acc_one = predict_price(1)
acc_two = predict_price(2)
acc_four = predict_price(4)

This is the predicted price based on the number of accomodates

In [23]:
print(acc_one)
print(acc_two)
print(acc_four)

78.8
126.0
197.6


## Evaluating Model Performance

We now have a function that can predict the price for any living space we want to list as long as we know the number of people it can accommodate. The function we wrote represents a machine learning model, which means that it outputs a prediction based on the input to the model.

A simple way to test the quality of your model is to:

1)   split the dataset into 2 partitions:

* the training set: contains the majority of the rows (75%)
* the test set: contains the remaining minority of the rows (25%)<br/>

2)   use the rows in the training set to predict the price value for the rows in the test set
* add new column named predicted_price to the test set

3)  compare the predicted_price values with the actual price values in the test set to see how accurate the predicted values were.

This validation process, where we use the training set to make predictions and the test set to predict values for, is known as **train/test validation**. Whenever you're performing machine learning, you want to perform validation of some kind to ensure that your machine learning model can make good predictions on new data.

While train/test validation isn't perfect, we'll use it to understand the validation process, to select an error metric, and then we'll dive into a more robust validation process

